File: android/app/src/main/java/com/bitcraps/app/sync/GameStateSynchronizer.kt
package com.bitcraps.app.sync

import android.util.Log
import com.google.gson.Gson
import kotlinx.coroutines.*
import kotlinx.coroutines.flow.*
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.atomic.AtomicLong
import kotlin.random.Random

/**
 * Real-time game state synchronization for cross-platform BitCraps gaming
 */
class GameStateSynchronizer(
    private val nodeId: String = generateNodeId()
) {
    companion object {
        private const val TAG = "GameStateSynchronizer"
        private const val SYNC_INTERVAL_MS = 100L
        private const val CONFLICT_RESOLUTION_TIMEOUT_MS = 5000L
        private const val MAX_NETWORK_LATENCY_MS = 2000L
        
        private fun generateNodeId(): String = "node_${Random.nextInt(10000)}"
    }
    
    private val gson = Gson()
    private val scope = CoroutineScope(Dispatchers.IO + SupervisorJob())
    
    // State management
    private val _gameState = MutableStateFlow(GameStateSnapshot.initial())
    val gameState: StateFlow<GameStateSnapshot> = _gameState.asStateFlow()
    
    private val _connectionState = MutableStateFlow(ConnectionState.DISCONNECTED)
    val connectionState: StateFlow<ConnectionState> = _connectionState.asStateFlow()
    
    private val _participants = MutableStateFlow<Map<String, ParticipantInfo>>(emptyMap())
    val participants: StateFlow<Map<String, ParticipantInfo>> = _participants.asStateFlow()
    
    // Synchronization internals
    private val vectorClock = VectorClock(nodeId)
    private val pendingOperations = ConcurrentHashMap<String, PendingOperation>()
    private val operationHistory = mutableListOf<GameOperation>()
    private val networkLatencyEstimate = AtomicLong(100L)
    
    // Network interface (to be implemented by actual transport layer)
    private var networkInterface: NetworkInterface? = null
    
    init {
        startSynchronizationLoop()
        startConflictResolutionLoop()
    }
    
    // Public API
    fun connect(networkInterface: NetworkInterface) {
        this.networkInterface = networkInterface
        networkInterface.setMessageHandler { message ->
            handleIncomingMessage(message)
        }
        _connectionState.value = ConnectionState.CONNECTED
        Log.d(TAG, "Connected to network with node ID: $nodeId")
    }
    
    fun disconnect() {
        networkInterface = null
        _connectionState.value = ConnectionState.DISCONNECTED
        _participants.value = emptyMap()
        Log.d(TAG, "Disconnected from network")
    }
    
    fun proposeGameOperation(operation: GameOperation) {
        Log.d(TAG, "Proposing operation: ${operation.type}")
        
        // Add vector clock timestamp
        val timestampedOperation = operation.copy(
            timestamp = vectorClock.increment(),
            nodeId = nodeId,
            id = generateOperationId()
        )
        
        // Apply optimistically for local responsiveness
        applyOperationLocally(timestampedOperation)
        
        // Add to pending operations for conflict resolution
        pendingOperations[timestampedOperation.id] = PendingOperation(
            operation = timestampedOperation,
            timestamp = System.currentTimeMillis(),
            acknowledged = setOf(nodeId)
        )
        
        // Broadcast to network
        broadcastOperation(timestampedOperation)
    }
    
    // Private implementation
    private fun startSynchronizationLoop() {
        scope.launch {
            while (isActive) {
                delay(SYNC_INTERVAL_MS)
                synchronizeWithPeers()
            }
        }
    }
    
    private fun startConflictResolutionLoop() {
        scope.launch {
            while (isActive) {
                delay(1000L)
                resolveConflicts()
                cleanupOldOperations()
            }
        }
    }
    
    private fun handleIncomingMessage(message: SyncMessage) {
        when (message.type) {
            MessageType.GAME_OPERATION -> handleGameOperation(message)
            MessageType.SYNC_REQUEST -> handleSyncRequest(message)
            MessageType.SYNC_RESPONSE -> handleSyncResponse(message)
            MessageType.HEARTBEAT -> handleHeartbeat(message)
            MessageType.CONFLICT_RESOLUTION -> handleConflictResolution(message)
        }
    }
    
    private fun handleGameOperation(message: SyncMessage) {
        val operation = gson.fromJson(message.payload, GameOperation::class.java)
        
        Log.d(TAG, "Received operation: ${operation.type} from ${operation.nodeId}")
        
        // Update vector clock
        vectorClock.update(operation.timestamp)
        
        // Check for conflicts
        val hasConflict = checkForConflicts(operation)
        
        if (hasConflict) {
            Log.w(TAG, "Conflict detected for operation: ${operation.id}")
            initiateConflictResolution(operation)
        } else {
            // Apply operation
            applyOperationLocally(operation)
            acknowledgeOperation(operation)
        }
    }
    
    private fun applyOperationLocally(operation: GameOperation) {
        val currentState = _gameState.value
        val newState = when (operation.type) {
            OperationType.DICE_ROLL -> {
                currentState.copy(
                    dice1 = operation.data["dice1"]?.toInt() ?: currentState.dice1,
                    dice2 = operation.data["dice2"]?.toInt() ?: currentState.dice2,
                    isRolling = operation.data["isRolling"]?.toBoolean() ?: false,
                    lastRollTimestamp = System.currentTimeMillis()
                )
            }
            OperationType.PLACE_BET -> {
                val playerId = operation.nodeId
                val amount = operation.data["amount"]?.toInt() ?: 0
                val newBets = currentState.playerBets.toMutableMap()
                newBets[playerId] = (newBets[playerId] ?: 0) + amount
                
                currentState.copy(
                    playerBets = newBets,
                    totalPot = currentState.totalPot + amount
                )
            }
            OperationType.PLAYER_JOIN -> {
                val playerId = operation.nodeId
                val playerName = operation.data["playerName"] ?: "Unknown"
                
                currentState.copy(
                    players = currentState.players + (playerId to PlayerInfo(
                        id = playerId,
                        name = playerName,
                        balance = 1000,
                        isConnected = true
                    ))
                )
            }
            OperationType.PLAYER_LEAVE -> {
                val playerId = operation.nodeId
                currentState.copy(
                    players = currentState.players - playerId,
                    playerBets = currentState.playerBets - playerId
                )
            }
            OperationType.GAME_PHASE_CHANGE -> {
                val newPhase = GamePhase.valueOf(operation.data["phase"] ?: "COME_OUT")
                val point = operation.data["point"]?.toIntOrNull()
                
                currentState.copy(
                    gamePhase = newPhase,
                    point = point
                )
            }
        }
        
        _gameState.value = newState.copy(
            version = vectorClock.getTime(),
            lastUpdateTimestamp = System.currentTimeMillis()
        )
        
        // Update operation history
        operationHistory.add(operation)
        if (operationHistory.size > 1000) {
            operationHistory.removeAt(0)
        }
        
        Log.d(TAG, "Applied operation: ${operation.type}, new state version: ${newState.version}")
    }
    
    private fun checkForConflicts(operation: GameOperation): Boolean {
        // Check if operation conflicts with pending operations
        return pendingOperations.values.any { pending ->
            pending.operation.timestamp.happensBefore(operation.timestamp) &&
            pending.operation.type == operation.type &&
            areOperationsConflicting(pending.operation, operation)
        }
    }
    
    private fun areOperationsConflicting(op1: GameOperation, op2: GameOperation): Boolean {
        return when {
            op1.type == OperationType.DICE_ROLL && op2.type == OperationType.DICE_ROLL -> {
                // Multiple simultaneous dice rolls conflict
                kotlin.math.abs((op1.data["timestamp"]?.toLong() ?: 0) - (op2.data["timestamp"]?.toLong() ?: 0)) < 1000
            }
            op1.type == OperationType.PLACE_BET && op2.type == OperationType.PLACE_BET -> {
                // Same player betting simultaneously
                op1.nodeId == op2.nodeId
            }
            else -> false
        }
    }
    
    private fun initiateConflictResolution(operation: GameOperation) {
        scope.launch {
            Log.d(TAG, "Initiating conflict resolution for operation: ${operation.id}")
            
            // Collect all conflicting operations
            val conflictingOps = pendingOperations.values
                .filter { areOperationsConflicting(it.operation, operation) }
                .map { it.operation }
                .plus(operation)
            
            // Resolve using deterministic ordering (timestamp + node ID)
            val resolvedOrder = conflictingOps.sortedWith(compareBy({ it.timestamp.toString() }, { it.nodeId }))
            val winningOperation = resolvedOrder.first()
            
            if (winningOperation.id == operation.id) {
                // This operation wins, apply it
                applyOperationLocally(operation)
                acknowledgeOperation(operation)
            } else {
                // Another operation wins, revert if necessary
                Log.d(TAG, "Operation ${operation.id} lost conflict resolution")
            }
            
            // Clean up conflicting pending operations
            conflictingOps.forEach { op ->
                if (op.id != winningOperation.id) {
                    pendingOperations.remove(op.id)
                }
            }
        }
    }
    
    private fun synchronizeWithPeers() {
        networkInterface?.let { network ->
            val syncMessage = SyncMessage(
                type = MessageType.SYNC_REQUEST,
                senderId = nodeId,
                payload = gson.toJson(SyncRequest(
                    currentVersion = vectorClock.getTime(),
                    lastOperationId = operationHistory.lastOrNull()?.id
                )),
                timestamp = System.currentTimeMillis()
            )
            
            network.broadcast(syncMessage)
        }
    }
    
    private fun broadcastOperation(operation: GameOperation) {
        networkInterface?.let { network ->
            val message = SyncMessage(
                type = MessageType.GAME_OPERATION,
                senderId = nodeId,
                payload = gson.toJson(operation),
                timestamp = System.currentTimeMillis()
            )
            
            network.broadcast(message)
        }
    }
    
    private fun acknowledgeOperation(operation: GameOperation) {
        pendingOperations[operation.id]?.let { pending ->
            val updatedPending = pending.copy(
                acknowledged = pending.acknowledged + nodeId
            )
            pendingOperations[operation.id] = updatedPending
            
            // If majority acknowledged, consider confirmed
            val totalParticipants = _participants.value.size + 1 // +1 for self
            if (updatedPending.acknowledged.size >= (totalParticipants + 1) / 2) {
                pendingOperations.remove(operation.id)
                Log.d(TAG, "Operation ${operation.id} confirmed by majority")
            }
        }
    }
    
    private fun resolveConflicts() {
        val now = System.currentTimeMillis()
        val timedOutOperations = pendingOperations.values.filter { 
            now - it.timestamp > CONFLICT_RESOLUTION_TIMEOUT_MS 
        }
        
        timedOutOperations.forEach { pending ->
            Log.w(TAG, "Operation ${pending.operation.id} timed out, applying with current state")
            pendingOperations.remove(pending.operation.id)
        }
    }
    
    private fun cleanupOldOperations() {
        // Keep only recent operations in history
        val cutoff = System.currentTimeMillis() - 60000L // 1 minute
        operationHistory.removeAll { it.data["timestamp"]?.toLong() ?: 0 < cutoff }
    }
    
    private fun handleSyncRequest(message: SyncMessage) {
        // TODO: Implement sync request handling
    }
    
    private fun handleSyncResponse(message: SyncMessage) {
        // TODO: Implement sync response handling
    }
    
    private fun handleHeartbeat(message: SyncMessage) {
        // Update participant info
        val participantInfo = ParticipantInfo(
            nodeId = message.senderId,
            lastSeen = System.currentTimeMillis(),
            latency = estimateLatency(message.timestamp)
        )
        
        _participants.value = _participants.value + (message.senderId to participantInfo)
    }
    
    private fun handleConflictResolution(message: SyncMessage) {
        // TODO: Implement conflict resolution message handling
    }
    
    private fun estimateLatency(messageTimestamp: Long): Long {
        val now = System.currentTimeMillis()
        val latency = now - messageTimestamp
        
        // Update running average
        val current = networkLatencyEstimate.get()
        val updated = (current * 0.9 + latency * 0.1).toLong()
        networkLatencyEstimate.set(updated)
        
        return latency
    }
    
    private fun generateOperationId(): String = "${nodeId}_${System.currentTimeMillis()}_${Random.nextInt(1000)}"
    
    fun cleanup() {
        scope.cancel()
        pendingOperations.clear()
        operationHistory.clear()
    }
}

// Data classes
data class GameStateSnapshot(
    val version: Map<String, Long>,
    val gamePhase: GamePhase,
    val dice1: Int,
    val dice2: Int,
    val isRolling: Boolean,
    val point: Int?,
    val players: Map<String, PlayerInfo>,
    val playerBets: Map<String, Int>,
    val totalPot: Int,
    val lastRollTimestamp: Long,
    val lastUpdateTimestamp: Long
) {
    companion object {
        fun initial() = GameStateSnapshot(
            version = emptyMap(),
            gamePhase = GamePhase.COME_OUT,
            dice1 = 0,
            dice2 = 0,
            isRolling = false,
            point = null,
            players = emptyMap(),
            playerBets = emptyMap(),
            totalPot = 0,
            lastRollTimestamp = 0L,
            lastUpdateTimestamp = System.currentTimeMillis()
        )
    }
}

data class GameOperation(
    val id: String,
    val type: OperationType,
    val nodeId: String,
    val timestamp: Map<String, Long>,
    val data: Map<String, String>
)

enum class OperationType {
    DICE_ROLL,
    PLACE_BET,
    PLAYER_JOIN,
    PLAYER_LEAVE,
    GAME_PHASE_CHANGE
}

enum class GamePhase {
    COME_OUT,
    POINT,
    GAME_OVER
}

data class PlayerInfo(
    val id: String,
    val name: String,
    val balance: Int,
    val isConnected: Boolean
)

data class PendingOperation(
    val operation: GameOperation,
    val timestamp: Long,
    val acknowledged: Set<String>
)

data class ParticipantInfo(
    val nodeId: String,
    val lastSeen: Long,
    val latency: Long
)

enum class ConnectionState {
    DISCONNECTED,
    CONNECTING,
    CONNECTED,
    ERROR
}

// Network interface (to be implemented by transport layer)
interface NetworkInterface {
    fun broadcast(message: SyncMessage)
    fun send(nodeId: String, message: SyncMessage)
    fun setMessageHandler(handler: (SyncMessage) -> Unit)
}

data class SyncMessage(
    val type: MessageType,
    val senderId: String,
    val payload: String,
    val timestamp: Long
)

enum class MessageType {
    GAME_OPERATION,
    SYNC_REQUEST,
    SYNC_RESPONSE,
    HEARTBEAT,
    CONFLICT_RESOLUTION
}

data class SyncRequest(
    val currentVersion: Map<String, Long>,
    val lastOperationId: String?
)

// Vector clock implementation for ordering events
class VectorClock(private val nodeId: String) {
    private val clock = mutableMapOf<String, Long>()
    
    init {
        clock[nodeId] = 0L
    }
    
    fun increment(): Map<String, Long> {
        clock[nodeId] = (clock[nodeId] ?: 0L) + 1L
        return clock.toMap()
    }
    
    fun update(otherClock: Map<String, Long>) {
        otherClock.forEach { (node, time) ->
            clock[node] = maxOf(clock[node] ?: 0L, time)
        }
        clock[nodeId] = (clock[nodeId] ?: 0L) + 1L
    }
    
    fun getTime(): Map<String, Long> = clock.toMap()
}

// Extension function for vector clock comparison
fun Map<String, Long>.happensBefore(other: Map<String, Long>): Boolean {
    var hasSmaller = false
    val allNodes = this.keys + other.keys
    
    for (node in allNodes) {
        val thisTime = this[node] ?: 0L
        val otherTime = other[node] ?: 0L
        
        if (thisTime > otherTime) return false
        if (thisTime < otherTime) hasSmaller = true
    }
    
    return hasSmaller
}

File: ios/BitCraps/Sources/Sync/GameStateSynchronizer.swift
import Foundation
import Combine
import os.log

/// Real-time game state synchronization for cross-platform BitCraps gaming
@available(iOS 15.0, *)
class GameStateSynchronizer: ObservableObject {
    private let logger = Logger(subsystem: "com.bitcraps.app", category: "GameStateSynchronizer")
    
    // Configuration
    private let nodeId: String
    private let syncInterval: TimeInterval = 0.1
    private let conflictResolutionTimeout: TimeInterval = 5.0
    private let maxNetworkLatency: TimeInterval = 2.0
    
    // Published state
    @Published var gameState: GameStateSnapshot
    @Published var connectionState: ConnectionState = .disconnected
    @Published var participants: [String: ParticipantInfo] = [:]
    
    // Internal state
    private var vectorClock: VectorClock
    private var pendingOperations: [String: PendingOperation] = [:]
    private var operationHistory: [GameOperation] = []
    private var networkLatencyEstimate: TimeInterval = 0.1
    
    // Network interface
    private var networkInterface: NetworkInterface?
    private var cancellables = Set<AnyCancellable>()
    private var syncTimer: Timer?
    private var conflictResolutionTimer: Timer?
    
    init(nodeId: String? = nil) {
        self.nodeId = nodeId ?? "node_\(Int.random(in: 1000...9999))"
        self.gameState = GameStateSnapshot.initial()
        self.vectorClock = VectorClock(nodeId: self.nodeId)
        
        setupTimers()
        logger.info("Initialized GameStateSynchronizer with nodeId: \(self.nodeId)")
    }
    
    deinit {
        cleanup()
    }
    
    // MARK: - Public API
    
    func connect(networkInterface: NetworkInterface) {
        self.networkInterface = networkInterface
        networkInterface.messageHandler = { [weak self] message in
            self?.handleIncomingMessage(message)
        }
        connectionState = .connected
        logger.info("Connected to network")
    }
    
    func disconnect() {
        networkInterface = nil
        connectionState = .disconnected
        participants.removeAll()
        logger.info("Disconnected from network")
    }
    
    func proposeGameOperation(_ operation: GameOperation) {
        logger.info("Proposing operation: \(operation.type)")
        
        // Add vector clock timestamp
        let timestampedOperation = GameOperation(
            id: generateOperationId(),
            type: operation.type,
            nodeId: nodeId,
            timestamp: vectorClock.increment(),
            data: operation.data
        )
        
        // Apply optimistically for local responsiveness
        applyOperationLocally(timestampedOperation)
        
        // Add to pending operations for conflict resolution
        pendingOperations[timestampedOperation.id] = PendingOperation(
            operation: timestampedOperation,
            timestamp: Date(),
            acknowledged: Set([nodeId])
        )
        
        // Broadcast to network
        broadcastOperation(timestampedOperation)
    }
    
    // MARK: - Private Implementation
    
    private func setupTimers() {
        // Synchronization timer
        syncTimer = Timer.scheduledTimer(withTimeInterval: syncInterval, repeats: true) { [weak self] _ in
            self?.synchronizeWithPeers()
        }
        
        // Conflict resolution timer
        conflictResolutionTimer = Timer.scheduledTimer(withTimeInterval: 1.0, repeats: true) { [weak self] _ in
            self?.resolveConflicts()
            self?.cleanupOldOperations()
        }
    }
    
    private func handleIncomingMessage(_ message: SyncMessage) {
        switch message.type {
        case .gameOperation:
            handleGameOperation(message)
        case .syncRequest:
            handleSyncRequest(message)
        case .syncResponse:
            handleSyncResponse(message)
        case .heartbeat:
            handleHeartbeat(message)
        case .conflictResolution:
            handleConflictResolution(message)
        }
    }
    
    private func handleGameOperation(_ message: SyncMessage) {
        guard let data = message.payload.data(using: .utf8),
              let operation = try? JSONDecoder().decode(GameOperation.self, from: data) else {
            logger.error("Failed to decode game operation")
            return
        }
        
        logger.info("Received operation: \(operation.type) from \(operation.nodeId)")
        
        // Update vector clock
        vectorClock.update(with: operation.timestamp)
        
        // Check for conflicts
        let hasConflict = checkForConflicts(operation)
        
        if hasConflict {
            logger.warning("Conflict detected for operation: \(operation.id)")
            initiateConflictResolution(for: operation)
        } else {
            // Apply operation
            applyOperationLocally(operation)
            acknowledgeOperation(operation)
        }
    }
    
    private func applyOperationLocally(_ operation: GameOperation) {
        let newState: GameStateSnapshot
        
        switch operation.type {
        case .diceRoll:
            newState = GameStateSnapshot(
                version: vectorClock.getTime(),
                gamePhase: gameState.gamePhase,
                dice1: Int(operation.data["dice1"] ?? "0") ?? gameState.dice1,
                dice2: Int(operation.data["dice2"] ?? "0") ?? gameState.dice2,
                isRolling: Bool(operation.data["isRolling"] ?? "false") ?? false,
                point: gameState.point,
                players: gameState.players,
                playerBets: gameState.playerBets,
                totalPot: gameState.totalPot,
                lastRollTimestamp: Date().timeIntervalSince1970,
                lastUpdateTimestamp: Date().timeIntervalSince1970
            )
            
        case .placeBet:
            let playerId = operation.nodeId
            let amount = Int(operation.data["amount"] ?? "0") ?? 0
            var newBets = gameState.playerBets
            newBets[playerId] = (newBets[playerId] ?? 0) + amount
            
            newState = GameStateSnapshot(
                version: vectorClock.getTime(),
                gamePhase: gameState.gamePhase,
                dice1: gameState.dice1,
                dice2: gameState.dice2,
                isRolling: gameState.isRolling,
                point: gameState.point,
                players: gameState.players,
                playerBets: newBets,
                totalPot: gameState.totalPot + amount,
                lastRollTimestamp: gameState.lastRollTimestamp,
                lastUpdateTimestamp: Date().timeIntervalSince1970
            )
            
        case .playerJoin:
            let playerId = operation.nodeId
            let playerName = operation.data["playerName"] ?? "Unknown"
            var newPlayers = gameState.players
            newPlayers[playerId] = PlayerInfo(
                id: playerId,
                name: playerName,
                balance: 1000,
                isConnected: true
            )
            
            newState = GameStateSnapshot(
                version: vectorClock.getTime(),
                gamePhase: gameState.gamePhase,
                dice1: gameState.dice1,
                dice2: gameState.dice2,
                isRolling: gameState.isRolling,
                point: gameState.point,
                players: newPlayers,
                playerBets: gameState.playerBets,
                totalPot: gameState.totalPot,
                lastRollTimestamp: gameState.lastRollTimestamp,
                lastUpdateTimestamp: Date().timeIntervalSince1970
            )
            
        case .playerLeave:
            let playerId = operation.nodeId
            var newPlayers = gameState.players
            var newBets = gameState.playerBets
            newPlayers.removeValue(forKey: playerId)
            newBets.removeValue(forKey: playerId)
            
            newState = GameStateSnapshot(
                version: vectorClock.getTime(),
                gamePhase: gameState.gamePhase,
                dice1: gameState.dice1,
                dice2: gameState.dice2,
                isRolling: gameState.isRolling,
                point: gameState.point,
                players: newPlayers,
                playerBets: newBets,
                totalPot: gameState.totalPot,
                lastRollTimestamp: gameState.lastRollTimestamp,
                lastUpdateTimestamp: Date().timeIntervalSince1970
            )
            
        case .gamePhaseChange:
            let phaseString = operation.data["phase"] ?? "comeOut"
            let newPhase = GamePhase(rawValue: phaseString) ?? .comeOut
            let point = Int(operation.data["point"] ?? "")
            
            newState = GameStateSnapshot(
                version: vectorClock.getTime(),
                gamePhase: newPhase,
                dice1: gameState.dice1,
                dice2: gameState.dice2,
                isRolling: gameState.isRolling,
                point: point,
                players: gameState.players,
                playerBets: gameState.playerBets,
                totalPot: gameState.totalPot,
                lastRollTimestamp: gameState.lastRollTimestamp,
                lastUpdateTimestamp: Date().timeIntervalSince1970
            )
        }
        
        DispatchQueue.main.async {
            self.gameState = newState
        }
        
        // Update operation history
        operationHistory.append(operation)
        if operationHistory.count > 1000 {
            operationHistory.removeFirst()
        }
        
        logger.info("Applied operation: \(operation.type), new state version: \(newState.version)")
    }
    
    private func checkForConflicts(_ operation: GameOperation) -> Bool {
        return pendingOperations.values.contains { pending in
            pending.operation.timestamp.happensBefore(operation.timestamp) &&
            pending.operation.type == operation.type &&
            areOperationsConflicting(pending.operation, operation)
        }
    }
    
    private func areOperationsConflicting(_ op1: GameOperation, _ op2: GameOperation) -> Bool {
        switch (op1.type, op2.type) {
        case (.diceRoll, .diceRoll):
            // Multiple simultaneous dice rolls conflict
            let timestamp1 = Double(op1.data["timestamp"] ?? "0") ?? 0
            let timestamp2 = Double(op2.data["timestamp"] ?? "0") ?? 0
            return abs(timestamp1 - timestamp2) < 1.0
            
        case (.placeBet, .placeBet):
            // Same player betting simultaneously
            return op1.nodeId == op2.nodeId
            
        default:
            return false
        }
    }
    
    private func initiateConflictResolution(for operation: GameOperation) {
        logger.info("Initiating conflict resolution for operation: \(operation.id)")
        
        // Collect all conflicting operations
        let conflictingOps = pendingOperations.values
            .filter { areOperationsConflicting($0.operation, operation) }
            .map { $0.operation } + [operation]
        
        // Resolve using deterministic ordering (timestamp + node ID)
        let resolvedOrder = conflictingOps.sorted { op1, op2 in
            if op1.timestamp.description != op2.timestamp.description {
                return op1.timestamp.description < op2.timestamp.description
            }
            return op1.nodeId < op2.nodeId
        }
        
        guard let winningOperation = resolvedOrder.first else { return }
        
        if winningOperation.id == operation.id {
            // This operation wins, apply it
            applyOperationLocally(operation)
            acknowledgeOperation(operation)
        } else {
            // Another operation wins
            logger.info("Operation \(operation.id) lost conflict resolution")
        }
        
        // Clean up conflicting pending operations
        for op in conflictingOps {
            if op.id != winningOperation.id {
                pendingOperations.removeValue(forKey: op.id)
            }
        }
    }
    
    private func synchronizeWithPeers() {
        guard let network = networkInterface else { return }
        
        let syncRequest = SyncRequest(
            currentVersion: vectorClock.getTime(),
            lastOperationId: operationHistory.last?.id
        )
        
        guard let data = try? JSONEncoder().encode(syncRequest),
              let payload = String(data: data, encoding: .utf8) else {
            return
        }
        
        let syncMessage = SyncMessage(
            type: .syncRequest,
            senderId: nodeId,
            payload: payload,
            timestamp: Date().timeIntervalSince1970
        )
        
        network.broadcast(syncMessage)
    }
    
    private func broadcastOperation(_ operation: GameOperation) {
        guard let network = networkInterface else { return }
        
        guard let data = try? JSONEncoder().encode(operation),
              let payload = String(data: data, encoding: .utf8) else {
            logger.error("Failed to encode operation for broadcast")
            return
        }
        
        let message = SyncMessage(
            type: .gameOperation,
            senderId: nodeId,
            payload: payload,
            timestamp: Date().timeIntervalSince1970
        )
        
        network.broadcast(message)
    }
    
    private func acknowledgeOperation(_ operation: GameOperation) {
        guard var pending = pendingOperations[operation.id] else { return }
        
        pending.acknowledged.insert(nodeId)
        pendingOperations[operation.id] = pending
        
        // If majority acknowledged, consider confirmed
        let totalParticipants = participants.count + 1 // +1 for self
        if pending.acknowledged.count >= (totalParticipants + 1) / 2 {
            pendingOperations.removeValue(forKey: operation.id)
            logger.info("Operation \(operation.id) confirmed by majority")
        }
    }
    
    private func resolveConflicts() {
        let now = Date()
        let timedOutOperations = pendingOperations.values.filter {
            now.timeIntervalSince($0.timestamp) > conflictResolutionTimeout
        }
        
        for pending in timedOutOperations {
            logger.warning("Operation \(pending.operation.id) timed out, applying with current state")
            pendingOperations.removeValue(forKey: pending.operation.id)
        }
    }
    
    private func cleanupOldOperations() {
        let cutoff = Date().timeIntervalSince1970 - 60.0 // 1 minute
        operationHistory.removeAll { operation in
            Double(operation.data["timestamp"] ?? "0") ?? 0 < cutoff
        }
    }
    
    private func handleSyncRequest(_ message: SyncMessage) {
        // TODO: Implement sync request handling
        logger.debug("Received sync request from \(message.senderId)")
    }
    
    private func handleSyncResponse(_ message: SyncMessage) {
        // TODO: Implement sync response handling
        logger.debug("Received sync response from \(message.senderId)")
    }
    
    private func handleHeartbeat(_ message: SyncMessage) {
        let participantInfo = ParticipantInfo(
            nodeId: message.senderId,
            lastSeen: Date().timeIntervalSince1970,
            latency: estimateLatency(messageTimestamp: message.timestamp)
        )
        
        DispatchQueue.main.async {
            self.participants[message.senderId] = participantInfo
        }
    }
    
    private func handleConflictResolution(_ message: SyncMessage) {
        // TODO: Implement conflict resolution message handling
        logger.debug("Received conflict resolution from \(message.senderId)")
    }
    
    private func estimateLatency(messageTimestamp: TimeInterval) -> TimeInterval {
        let now = Date().timeIntervalSince1970
        let latency = now - messageTimestamp
        
        // Update running average
        networkLatencyEstimate = networkLatencyEstimate * 0.9 + latency * 0.1
        
        return latency
    }
    
    private func generateOperationId() -> String {
        return "\(nodeId)_\(Int(Date().timeIntervalSince1970 * 1000))_\(Int.random(in: 100...999))"
    }
    
    private func cleanup() {
        syncTimer?.invalidate()
        conflictResolutionTimer?.invalidate()
        cancellables.removeAll()
        pendingOperations.removeAll()
        operationHistory.removeAll()
    }
}

// MARK: - Data Models

struct GameStateSnapshot: Codable {
    let version: [String: Int64]
    let gamePhase: GamePhase
    let dice1: Int
    let dice2: Int
    let isRolling: Bool
    let point: Int?
    let players: [String: PlayerInfo]
    let playerBets: [String: Int]
    let totalPot: Int
    let lastRollTimestamp: TimeInterval
    let lastUpdateTimestamp: TimeInterval
    
    static func initial() -> GameStateSnapshot {
        return GameStateSnapshot(
            version: [:],
            gamePhase: .comeOut,
            dice1: 0,
            dice2: 0,
            isRolling: false,
            point: nil,
            players: [:],
            playerBets: [:],
            totalPot: 0,
            lastRollTimestamp: 0,
            lastUpdateTimestamp: Date().timeIntervalSince1970
        )
    }
}

struct GameOperation: Codable {
    let id: String
    let type: OperationType
    let nodeId: String
    let timestamp: [String: Int64]
    let data: [String: String]
}

enum OperationType: String, Codable {
    case diceRoll = "diceRoll"
    case placeBet = "placeBet"
    case playerJoin = "playerJoin"
    case playerLeave = "playerLeave"
    case gamePhaseChange = "gamePhaseChange"
}

enum GamePhase: String, Codable {
    case comeOut = "comeOut"
    case point = "point"
    case gameOver = "gameOver"
}

struct PlayerInfo: Codable {
    let id: String
    let name: String
    let balance: Int
    let isConnected: Bool
}

struct PendingOperation {
    let operation: GameOperation
    let timestamp: Date
    var acknowledged: Set<String>
}

struct ParticipantInfo {
    let nodeId: String
    let lastSeen: TimeInterval
    let latency: TimeInterval
}

enum ConnectionState {
    case disconnected
    case connecting
    case connected
    case error
}

// MARK: - Network Interface

protocol NetworkInterface {
    var messageHandler: ((SyncMessage) -> Void)? { get set }
    func broadcast(_ message: SyncMessage)
    func send(to nodeId: String, message: SyncMessage)
}

struct SyncMessage: Codable {
    let type: MessageType
    let senderId: String
    let payload: String
    let timestamp: TimeInterval
}

enum MessageType: String, Codable {
    case gameOperation = "gameOperation"
    case syncRequest = "syncRequest"
    case syncResponse = "syncResponse"
    case heartbeat = "heartbeat"
    case conflictResolution = "conflictResolution"
}

struct SyncRequest: Codable {
    let currentVersion: [String: Int64]
    let lastOperationId: String?
}

// MARK: - Vector Clock

class VectorClock {
    private let nodeId: String
    private var clock: [String: Int64] = [:]
    
    init(nodeId: String) {
        self.nodeId = nodeId
        clock[nodeId] = 0
    }
    
    func increment() -> [String: Int64] {
        clock[nodeId] = (clock[nodeId] ?? 0) + 1
        return clock
    }
    
    func update(with otherClock: [String: Int64]) {
        for (node, time) in otherClock {
            clock[node] = max(clock[node] ?? 0, time)
        }
        clock[nodeId] = (clock[nodeId] ?? 0) + 1
    }
    
    func getTime() -> [String: Int64] {
        return clock
    }
}

// MARK: - Extensions

extension Dictionary where Key == String, Value == Int64 {
    func happensBefore(_ other: [String: Int64]) -> Bool {
        var hasSmaller = false
        let allNodes = Set(self.keys).union(Set(other.keys))
        
        for node in allNodes {
            let thisTime = self[node] ?? 0
            let otherTime = other[node] ?? 0
            
            if thisTime > otherTime { return false }
            if thisTime < otherTime { hasSmaller = true }
        }
        
        return hasSmaller
    }
}